core/primitives/address.cpp

```diff
@@ -5,6 +5,163 @@
+                          [](const BLSPublicKeyHash &v) { return Buffer(v.begin(), v.end()); });
+  }
+
+  uint64_t Address::Value() {
+    uint64_t *ret_ptr = boost::get<uint64_t>(&data_);
+    return ret_ptr == nullptr ? 0 : *ret_ptr;
+  }
+
+  std::string Address::String() {
+    using base32 = cppcodec::base32_rfc4648;
+
+    std::string res{};
+
+    char networkPrefix = 't';
+    if (network_ == Network::testnet) networkPrefix = 'f';
+    res.push_back(networkPrefix);
+
+    AddressProtocol p = this->Protocol();
+    res.append(std::to_string(to_underlying(p)));
+
+    if (p == AddressProtocol::ID) {
+      return res.append(std::to_string(this->Value()));
+    }
+
+    Buffer databytes = this->SerializeData();
+
+    Buffer ingest{};
+    ingest.putUint8(static_cast<uint8_t>(to_underlying(p)));
+    ingest.putBuffer(databytes);
+    Blob<4> chksum = Checksum(ingest);
+
+    databytes.put(chksum);
+    auto encoded = base32::encode(databytes);
+    return res.append(encoded);
+  }
+
+  outcome::result<std::unique_ptr<Address> > makeAddress(Network net, uint8_t protocol, const Buffer &payload) {
+    if (!(net == Network::testnet || net == Network::mainnet)) {
+      return outcome::failure(AddressError::UNKNOWN_NETWORK);
+    }
+    switch (protocol) {
+      case to_underlying(AddressProtocol::ID): {
+        auto decoded = codec::leb128::decode<uint64_t>(payload.toVector());
+        if (!decoded) {
+          return outcome::failure(AddressError::INVALID_PAYLOAD);
+        }
+        return outcome::success(std::unique_ptr<Address>(new Address(net, decoded.value())));
+      }
+      case to_underlying(AddressProtocol::SECP256K1): {
+        if (payload.size() != kBlake2b160HashSize) {
+          return outcome::failure(AddressError::INVALID_PAYLOAD);
+        }
+        Secp256k1PublicKeyHash hash{};
+        std::copy_n(payload.begin(), kBlake2b160HashSize, hash.begin());
+        return outcome::success(std::unique_ptr<Address>(new Address(net, hash)));
+      }
+      case to_underlying(AddressProtocol::Actor): {
+        if (payload.size() != kBlake2b160HashSize) {
+          return outcome::failure(AddressError::INVALID_PAYLOAD);
+        }
+        ActorExecHash hash{};
+        std::copy_n(payload.begin(), kBlake2b160HashSize, hash.begin());
+        return outcome::success(std::unique_ptr<Address>(new Address(net, hash)));
+      }
+      case to_underlying(AddressProtocol::BLS): {
+        if (payload.size() != kBLSHashSize) {
+          return outcome::failure(AddressError::INVALID_PAYLOAD);
+        }
+        BLSPublicKeyHash hash{};
+        std::copy_n(payload.begin(), kBLSHashSize, hash.begin());
+        return outcome::success(std::unique_ptr<Address>(new Address(net, hash)));
+      }
+      default:
+        return outcome::failure(AddressError::UNKNOWN_PROTOCOL);
+    }
+  }
+
+  bool operator==(const Address &lhs, const Address &rhs) {
```

Comment:
Should we also check protocol? I understand that the same hash is almost impossible, but
f2abc == f3abc

Reply:
An idea like that has come across my mind, of course but what are the odds? Anyway, checking the protocol is a small overhead after all, so I'll just add it